insulin_on_board.py

```diff
@@ -0,0 +1,104 @@
+        Returns a list of time-basal lists 
+    """
+    time_vals = []
+    for basal_entry in basal_data:
+        if basal_entry['type'] == 'basal' and basal_entry['deliveryType'] != 'suspend':
+            rate = basal_entry["rate"] #unit per hour
+            duration = basal_entry["duration"]/1000/60 #in minutes
+            str_time = basal_entry["deviceTime"]
+            start_time = datetime.strptime(str_time, '%Y-%m-%dT%H:%M:%S')
+            total_insulin_delivered = rate * (duration/60) 
+            num_segments = duration / 5 #5 minute segments 
+            insulin_per_segment = total_insulin_delivered / num_segments
+        next_time = int(start_time.strftime('%s')) #in seconds
+        end_date = start_time + timedelta(minutes=duration)
+        end_time = int(end_date.strftime('%s'))
+        while next_time < end_time:
+            time_vals.append([next_time, insulin_per_segment])
+            next_time += 5 * 60 #next time -- 5 minutes later (in seconds)  
+    return time_vals
+
+def format_bolus_for_wizard(bolus_data):
+    """ Retrieve rates, times and duration values from bolus data to generate IOB values
+        Returns a list of time-bolus lists
+    """ 
+    time_vals = []
+    for bolus_entry in bolus_data:
+        str_time = bolus_entry["time"]
+        start_time = datetime.strptime(str_time, '%Y-%m-%dT%H:%M:%S.000Z')
+        if bolus_entry['subType'] == "normal":
+            initial_insulin = bolus_entry["normal"]
+            time_vals.append([int(start_time.strftime('%s')), initial_insulin])
+        else:
+            duration = bolus_entry["duration"]/1000/60 #in minutes
+            if duration > 0:
+                num_segments = duration / 5 #5 minute segments 
+                extended_insulin = bolus_entry["extended"]
+                insulin_per_segment = extended_insulin / num_segments
+                if bolus_entry['subType'] == "dual/square":
+                    initial_insulin = bolus_entry["normal"]
+                    time_vals.append([int(start_time.strftime('%s')), initial_insulin])
+                next_time = int(start_time.strftime('%s')) #in seconds
+                end_date = start_time + timedelta(minutes=duration)
+                end_time = int(end_date.strftime('%s'))
+                while next_time < end_time:
+                    time_vals.append([next_time, insulin_per_segment])
+                    next_time += 5 * 60 #next time -- 5 minutes later (in seconds)  
+    return time_vals            
+        
+def creare_iob_dict(bolus_data, action_time):
+    """ Return a dictionary with insulin on board values for every timestamp in bolus_data"""
+    time_vals = format_bolus_for_wizard(bolus_data)
+    iob_dict = {}
+    for time_bolus in time_vals:
+        remaining_time = action_time * 60 #in minutes
+        step = 0 
+        time = time_bolus[0]
+        initial_value = time_bolus[1]
```

Comment:
time_vals is a list of time-value lists. each time value (what i called time_bolus) has as its first entry an epoch timestamp at [0] and an insulin value at [1]. I'll try to clarify it in the code as well